### PR TITLE
Cmd line

### PR DESCRIPTION
**Pull request by jlchang**
Um, my first pull request - all feedback is appreciated!

one request, please take a look at the attribution to the Seurat tutorial that was written for the "process_data" function. Happy to make a more explicit/expansive acknowledgement or follow a style guide.

### CODE REVIEW

File: RF_PBMC.R [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+suppressPackageStartupMessages(library(logging))
+suppressPackageStartupMessages(library(dplyr))
+suppressPackageStartupMessages(library(randomForest))
+suppressPackageStartupMessages(library(Seurat))
+
+ 
+# process_data function follows Seurat tutorial for processing PBMC data set made available by 10x Genomics
+# dataset https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz
+# Seurat tutorial https://satijalab.org/seurat/pbmc3k_tutorial.html
+process_data <- function(input.data){
```
Comment (TimothyTickle) [inline RF_PBMC.R]: This function can used more comments on what is happening internally.

File: RF_PBMC.R [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	dense.size <- object.size(x = as.matrix(x = input.data))
+	print(dense.size)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: Dense size can be removed you are not using it in the rest of the code. Remove lines 16 and 17. This is a point in the tutorials about how different representations in memory of the data are more or less efficient.

File: RF_PBMC.R [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	seuratObj <- CreateSeuratObject(raw.data = input.data, min.cells = 3, min.genes = 200, project = "10X_data")
+	print(seuratObj)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: min.cells and min.genes can be optional arguments.

Comment (TimothyTickle) [reply RF_PBMC.R]: move print statements to logging using the logging class.

File: RF_PBMC.R [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	mito.genes <- grep(pattern = "^MT-", x = rownames(x = seuratObj@data), value = TRUE)
+	percent.mito <- Matrix::colSums(seuratObj@raw.data[mito.genes, ])/Matrix::colSums(seuratObj@raw.data)
+	seuratObj <- AddMetaData(object = seuratObj, metadata = percent.mito, col.name = "percent.mito")
+	vln <- VlnPlot(object = seuratObj, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)
+	print(vln)
+	par(mfrow = c(1, 2))
+	GenePlot(object = seuratObj, gene1 = "nUMI", gene2 = "percent.mito")
+	GenePlot(object = seuratObj, gene1 = "nUMI", gene2 = "nGene")
+	seuratObj <- FilterCells(object = seuratObj, subset.names = c("nGene", "percent.mito"), low.thresholds = c(200, -Inf), high.thresholds = c(2500, 0.05))
```
Comment (TimothyTickle) [inline RF_PBMC.R]: lines 20 - 28 could be a separate block of code for filtering. Thinking how to reduce this large chunk of code.

Comment (TimothyTickle) [reply RF_PBMC.R]: FilterCells can be parameterized as well. For now you can parameterize low.thresholds and high.thresholds into 4 arguments.

File: RF_PBMC.R [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	seuratObj <- NormalizeData(object = seuratObj, normalization.method = "LogNormalize", scale.factor = 10000)
+	seuratObj <- FindVariableGenes(object = seuratObj, mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: x.low.cutoff, x.high.cuttoff, and y.cuttoff should be optional arguments

File: RF_PBMC.R [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	length(x = seuratObj@var.genes)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: Remove this line or print out the genes in logging. Length is not being used.

File: RF_PBMC.R [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	seuratObj <- ScaleData(object = seuratObj, vars.to.regress = c("nUMI", "percent.mito"))
+	seuratObj <- RunPCA(object = seuratObj, pc.genes = seuratObj@var.genes, do.print = TRUE, pcs.print = 1:5, genes.print = 5)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: pcs.print and genes.print can be optional arguments

File: RF_PBMC.R [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	PrintPCA(object = seuratObj, pcs.print = 1:5, genes.print = 5, use.full = FALSE)
+	viz <- VizPCA(object = seuratObj, pcs.use = 1:2)
+	print(viz)
+	pca <- PCAPlot(object = seuratObj, dim.1 = 1, dim.2 = 2)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: 34 -37 are printing/plotting about principle components derived from PCA. Make sure pcs.print in the Run PCA and the above functions for PCs are coordinated. So pcs.print in PrintPCA should print the same PCs as Run. VIzPCA should also viz the same PCs and PCAPlot should print each of the PCs from pcs.print at least once as one of the dimensions in one of the plots. You do not have to do the full combinations of 1-5 choose 2 to plot in the X-Y of the plot but you can for instance plot 1v2, 2v3, 3v4, 4v5 and get mostly what people will want with the PCAPlot functionality.

Comment (TimothyTickle) [reply RF_PBMC.R]: All the PCA functionality could also be a separate function

File: RF_PBMC.R [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+	seuratObj <- ProjectPCA(object = seuratObj, do.print = FALSE)
+	seuratObj <- JackStraw(object = seuratObj, num.replicate = 100, display.progress = FALSE)
+	JackStrawPlot(object = seuratObj, PCs = 1:12)
+	jackstraw <- JackStrawPlot(object = seuratObj, PCs = 1:12)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: will need to make the JackStraw PCs an optional argument

File: RF_PBMC.R [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+	print(jackstraw)
+	elbow <- PCElbowPlot(object = seuratObj)
+	print(elbow)
+	seuratObj <- FindClusters(object = seuratObj, reduction.type = "pca", dims.use = 1:10, resolution = 0.6, print.output = 0, save.SNN = TRUE)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: dims.use should be an optional argument.

File: RF_PBMC.R [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+	PrintFindClustersParams(object = seuratObj)
+	seuratObj <- RunTSNE(object = seuratObj, dims.use = 1:10, do.fast = TRUE)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: dims.use should be an argument and should be the same as FindClusters?

File: RF_PBMC.R [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+	TSNEPlot(object = seuratObj)
+	# find all markers of cluster 1
+	cluster1.markers <- FindMarkers(object = seuratObj, ident.1 = 1, min.pct = 0.25)
+	print(x = head(x = cluster1.markers, n = 5))
+	# find all markers distinguishing cluster 5 from clusters 0 and 3
+	cluster5.markers <- FindMarkers(object = seuratObj, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: Why are these particular clusters selected in lines 50-53?

File: RF_PBMC.R [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+	print(x = head(x = cluster5.markers, n = 5))
+	# find markers for every cluster compared to all remaining cells, report
+	# only the positive ones
+	seuratObj.markers <- FindAllMarkers(object = seuratObj, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
+	seuratObj.markers %>% group_by(cluster) %>% top_n(2, avg_logFC)
+	cluster1.markers <- FindMarkers(object = seuratObj, ident.1 = 0, thresh.use = 0.25, test.use = "roc", only.pos = TRUE)
```
Comment (TimothyTickle) [inline RF_PBMC.R]: thresh.use should be an optional argument.

File: RF_PBMC.R [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+# save in tsv file - gene names for top 100 factors from the classification model in order of importance
+classify_data <- function(seuratObj){
+	classifier_input_x = t(seuratObj@data)
+	print(classifier_input_x)
+	classifier_input_y = as.factor(as.numeric(seuratObj@ident))
+	trained_classifier = randomForest(x = as.matrix(classifier_input_x) , y=factor(classifier_input_y) , importance = TRUE )
+	print(trained_classifier) 
+	ranked_list = trained_classifier$importance[order(-trained_classifier$importance[,'MeanDecreaseAccuracy']),]
+	dput(rownames(ranked_list)[1:300]) 
+	FINAL_LIST = rownames(ranked_list)[1:100]
```
Comment (TimothyTickle) [inline RF_PBMC.R]: line 73 and 74 should use optional parameters for the subsetting ranges.

File: RF_PBMC.R [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+                                            "--input_dir directory"))
+
+
+# input (string) path to Cell Ranger output directory
+pargs <- optparse::add_option(pargs, c("--input_dir"),
+                              type="character",
+                              action="store",
+                              dest="input_dir",
+                              metavar="Input_Directory",
+                              help=paste("Input directory for analysis.",
```
Comment (TimothyTickle) [inline RF_PBMC.R]: this is a 10x output directory, include that information in the description.

File: RF_PBMC.R [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+args_parsed <- optparse::parse_args(pargs)
+
+
+
+# Make sure the input directory exists
+if(!file.exists(args_parsed$input_dir[1])){
+    error_message <- paste0("Provided input directory, ",
+							args_parsed$input_dir[1], ",  does not exist.",
+    						"Please check your input and try again")
```
Comment (TimothyTickle) [inline RF_PBMC.R]: Make spacing consistent.in 103-105. If you are using tabs please do not use tabs in any code. Use other spacing.

File: RF_PBMC.R [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    stop(error_message)
+}
+
+
+# Manage inputs
```
Comment (TimothyTickle) [inline RF_PBMC.R]: please add more of a comment here.

**Review by TimothyTickle — CHANGES_REQUESTED** [review]
Did a first pass on the review. Mostly cosmetic for ease of use and maintenance :-)

File: RF_PBMC.R (hunk 2 above)
Comment (jlchang) [reply RF_PBMC.R]: removed

File: RF_PBMC.R (hunk 1 above)
Comment (jlchang) [reply RF_PBMC.R]: description of implementation per Seurat tutorial site added as comment to describe the implementation

File: RF_PBMC.R (hunk 3 above)
Comment (jlchang) [reply RF_PBMC.R]: Comment added indicating TODO for parameterization
Comment added setting TODO for moving from print statements to logging

File: RF_PBMC.R (hunk 4 above)
Comment (jlchang) [reply RF_PBMC.R]: Comment added indicating TODO for parameterization
Does "separate block of code" mean a separate function or just visual separation (whitespace + comments)?

File: RF_PBMC.R (hunk 5 above)
Comment (jlchang) [reply RF_PBMC.R]: Comment added indicating TODO for parameterization

File: RF_PBMC.R (hunk 6 above)
Comment (jlchang) [reply RF_PBMC.R]: removed unneeded line

File: RF_PBMC.R (hunk 7 above)
Comment (jlchang) [reply RF_PBMC.R]: Comment added indicating TODO for parameterization

File: RF_PBMC.R (hunk 8 above)
Comment (jlchang) [reply RF_PBMC.R]: Indicated in comment to move PCA commands to new function

File: RF_PBMC.R (hunk 9 above)
Comment (jlchang) [reply RF_PBMC.R]: Comment added indicating TODO for parameterization
for JackStraw function num.replicate = 100, should num.replicate also be parameterized?

File: RF_PBMC.R (hunk 10 above)
Comment (jlchang) [reply RF_PBMC.R]: Comment added indicating TODO for parameterization

File: RF_PBMC.R (hunk 11 above)
Comment (jlchang) [reply RF_PBMC.R]: Comment indicating same value should be used for both functions

File: RF_PBMC.R (hunk 12 above)
Comment (jlchang) [reply RF_PBMC.R]: Will find out - added TODO to either comment on rationale and/or parameterize

File: RF_PBMC.R (hunk 13 above)
Comment (jlchang) [reply RF_PBMC.R]: Comment added indicating TODO for parameterization

File: RF_PBMC.R (hunk 14 above)
Comment (jlchang) [reply RF_PBMC.R]: Comment added indicating TODO for parameterization

File: RF_PBMC.R (hunk 15 above)
Comment (jlchang) [reply RF_PBMC.R]: Updated description of input to reflect expected format

File: RF_PBMC.R (hunk 17 above)
Comment (jlchang) [reply RF_PBMC.R]: comment updated to be more descriptive

File: RF_PBMC.R (hunk 16 above)
Comment (jlchang) [reply RF_PBMC.R]: will clean up tabs and use spaces instead

File: RF_PBMC.R (hunk 12 above)
Comment (jlchang) [reply RF_PBMC.R]: demonstration code, removing to leave just FindAllMarkers

**Pull request closed without merging**